### ISSUE TITLE
Slow rsync  on OmniOS r151026

### ISSUE THREAD

**Issue by priyadarshan** [omniosorg/omnios-build#820]
We have been measuring consistently slow io with rsync as installed from IPS Packages for OmniOS r151026.

In some measurement, slow enough to not saturate a mirror vdev write-speed (slow, prosumer-grade WD RED, with about 240M write throughput).

Trying with a different rsync, installed from [Joyent's pkgsrc](https://github.com/joyent/pkgsrc/tree/trunk/net/rsync), speed is able to saturate that vdev (also comparable with FreeBSD).

Tests were also done on local SSD (no network involved), using a 10G file,

```
root@omniosce:~# dd if=/dev/urandom of=random.bin bs=1048576 count=10240
```

1. rsync to same SDD (rsync from OmniOS r151026 IPS packages),

```
root@omniosce:~# time rsync -a random.bin random2.bin

real    1m25.644s
user    2m24.261s
sys     0m14.273s
```

2. rsync to same SDD (rsync from Joyent's pkgsrc),

```
root@omniosce:~# time /opt/local/bin/rsync -a random.bin random1.bin

real    0m31.302s
user    0m40.634s
sys     0m13.994s
```

**Comment by citrus-it** [comment]
Can you try this updated version?

`pkg apply-hot-fix https://downloads.omniosce.org/pkg/r151026/820.p5p`

This one is built 64-bit, which is the only difference I can see between the old one and Joyent's

**Comment by priyadarshan** [comment]
Thank you, I have installed new version:
```
root@omniosce:~# ls -l /usr/bin/rsync
-rwxr-xr-x 1 root bin 687392 Jun 15 09:33 /usr/bin/rsync
```

As a side note, filesizes are a bit different:
```
root@omniosce:~# ls -l /opt/local/bin/rsync
-rwxr-xr-x 1 root root 627168 Apr  7 16:41 /opt/local/bin/rsync
```

I have timed new version, with basically same results:

```
root@omniosce:~# time /usr/bin/rsync -a random.bin random2.bin

real    1m26.651s
user    2m24.479s
sys     0m15.525s
```

Looking at `zpool iostat` as a very rudimentary io measure, I can see neither of those rsync saturate the SSD.

OmniOS version is about 130 MB/sec, and `pkgsrc` version is about 340 MB/sec.

Testing with `cp` I get  about 1.5 GB/sec, :

```
root@omniosce:~# time cp random.bin random1.bin

real    0m5.653s
user    0m0.131s
sys     0m4.999s
```

I alzo tested with an large text file (that is, quite compressible in respect to the random file above), with comparable results.

Could the bottleneck be in reading, instead of writing? Is there some tweak to do somewhere, perhaps bound to the SSD device driver? But again, that would not explain why would `cp` be so much faster.

**Comment by citrus-it** [comment]
Whatever is going on here, the difference is in userspace. The time spent in read() and write() is pretty consistent.
```
omni# truss -c /opt/local/bin/rsync -a random.bin random7.bin

syscall               seconds   calls  errors
read                    9.653   40970
write                   6.012  327685
pollsys                 2.421  327691
                     --------  ------   ----
sys totals:            18.090  696545     31
usr time:              32.415
elapsed:               73.400


omni# truss -c rsync -a random.bin random8.bin

syscall               seconds   calls  errors
read                    9.849   40967
write                   5.893  327685
pollsys                 2.259  327691
                     --------  ------   ----
sys totals:            18.003  696478      9
usr time:             107.377
elapsed:              147.930

```

**Comment by citrus-it** [comment]
With checksumming turned off, performance is identical:

```
omni# ptime rsync -a --checksum-choice=none random.bin random1.bin

real       19.801894330
user        6.056084813
sys        19.281762574

omni# ptime /opt/local/bin/rsync -a --checksum-choice=none random.bin random2.bin

real       19.665593874
user        6.117851191
sys        19.391562651
```

**Comment by priyadarshan** [comment]
Ah, that makes sense! Somehow the two versions use a different checksumming.

On my system, the OmniOS version is even a bit faster.

```
root@omniosce:~# ptime /usr/bin/rsync --checksum-choice=none random.bin random1.bin

real       12.151269438
user        5.758612956
sys        11.848265348

root@itstore:~# rm random1.bin

root@omniosce:~# ptime /opt/local/bin/rsync --checksum-choice=none random.bin random1.bin

real       14.775809445
user        5.557372669
sys        14.644037132
```

Thank you, that explains it perfectly. You have no idea how much I digegd and searched about that "mystery".

Now I can proceed moving several terabytes from one server to our first OmniOS server.

This ticket can be close as far as it concerns me.

Thank you for considering this, your tiome and for showing how to use `truss`, and about `ptime`, too. Very useful.

**Comment by citrus-it** [comment]
Can you try `pkg apply-hot-fix https://downloads.omniosce.org/pkg/r151026/820.p5p` again please?

**Comment by priyadarshan** [comment]
Thank you for hot-fix, now they are basically identical.

```
root@omniosce:~# ptime /usr/bin/rsync  random.bin random1.bin

real       30.572892048
user       42.527609387
sys        12.096188903

root@itstore:~# rm random2.bin

root@omniosce:~# ptime /opt/local/bin/rsync  random.bin random2.bin

real       31.418166438
user       40.713462591
sys        14.195722083
```

Long live OmniOS!

Thank you.

**Comment by citrus-it** [comment]
Great, thanks for reporting it.

I've raised a PR to fix this in r151026 - #828  - the updated rsync package will be in next week's release.
